Codec/Archive/Tar/Check.hs

```diff
@@ -65,24 +65,46 @@ checkSecurity = checkEntries checkEntrySecurity
 checkEntrySecurity :: Entry -> Maybe FileNameError
 checkEntrySecurity entry = case entryContent entry of
     HardLink     link -> check (entryPath entry)
-                 `mplus` check (fromLinkTarget link)
+                 `mplus` checkLink (entryPath entry) (fromLinkTarget link)
     SymbolicLink link -> check (entryPath entry)
-                 `mplus` check (fromLinkTarget link)
+                 `mplus` checkLink (entryPath entry) (fromLinkTarget link)
     _                 -> check (entryPath entry)
 
   where
+
+    checkCommon name =
+        FilePath.Native.isAbsolute name || not (FilePath.Native.isValid name)
+
     check name
-      | FilePath.Native.isAbsolute name
+      | checkCommon name
       = Just $ AbsoluteFileName name
 
-      | not (FilePath.Native.isValid name)
+      | any (=="..") (FilePath.Native.splitDirectories name)
       = Just $ InvalidFileName name
 
-      | any (=="..") (FilePath.Native.splitDirectories name)
+      | otherwise = Nothing
+
+    -- checkLink introduced to handle https://github.com/haskell/tar/issues/32
+    checkLink name link
+      | checkCommon name
+      = Just $ AbsoluteFileName name
+
+      | linkDepth link name > 0
```

Comment:
Doesn't this mean "the link target never leaves the directory of the symlink itself" as opposed to "the link target never leaves the base directory of the tar archive"?